### PR TITLE
Refactor Platform_detection.go

### PR DESCRIPTION
**Pull request by crombus**
Signed-off-by: crombus <pkundra@redhat.com>

### CODE REVIEW

File: pkg/controller/storagecluster/cephblockpools_test.go
Comment (obnoxxx) [inline pkg/controller/storagecluster/cephblockpools_test.go:35]: why is this hunk moved?

File: pkg/controller/storagecluster/cephfilesystem_test.go
Comment (obnoxxx) [inline pkg/controller/storagecluster/cephfilesystem_test.go:34]: why is this hunk moved?

**Review by obnoxxx — COMMENTED** [review]
Hi @crombus - thanks for working on the refactoring!
When we discussed it earlier today, I had imagined to approach it a bit different.

* Writing a wrapper around current getplatform and avoidobjectstore instead of changing current avoidobjectstore, we can keep the diff smaller and avoid having to change the test code immediately.
* If we let the wrapper hang off of `ReconcileStorageCluster`, we can avoid the extra argument.

Here are a few raw patches to illustrate what I meant: https://github.com/obnoxxx/ocs-operator/commits/refactor-avoid-object-store

When thinking about it initially, I had thought of just having a bool return, but we don't want to ignore the error, so the callers don't get simplified significantly... Therefore it is possibly not really worth the hassle. But I just hate having to repeat the same pattern call GetPlatform, then call avoidObjectStore in multiple locations, that's what triggered this, so it may not get much shorter, but at least less error-prone due to less repetition.

On top of such changes one could start using the wrapper in the tests as well, but at least it's not necessary to change all test code right away.

**Review by jarrpa — APPROVED** [review]
Overall this looks good. If you could do one more rebase, I'll LGTM it.

**Review by umangachapagain — COMMENTED** [review]
I am trying to understand how this re-factor is helpful or better?

The way I see it, we still need the `platform` value in the caller for this newly added wrapper.
So why not just `GetPlatform()` first and then call `AvoidObjectStore(platform)` ?

This wrapper seems to hide the fact that we aren't actually avoiding ObjectStore but instead trying to decide if we should, based on the platform we are running on.
The naming makes it worse. When I call `AvoidObjectStore()` I expect it to avoid object store and tell me if it was or wasn't able to do so rather than telling me if I can do it and on what platform am I running.

/hold

**Comment by jarrpa** [comment]
> I am trying to understand how this re-factor is helpful or better?
> 
> The way I see it, we still need the `platform` value in the caller for this newly added wrapper.
> So why not just `GetPlatform()` first and then call `AvoidObjectStore(platform)` ?

For the purposes of determining if we will deploy RGW, we have no need to know what the platform actually is. This method is way more streamlined. One function call to get both the boolean and platform (if applicable) makes more sense to me.

> This wrapper seems to hide the fact that we aren't actually avoiding ObjectStore but instead trying to decide if we should, based on the platform we are running on.
> The naming makes it worse. When I call `AvoidObjectStore()` I expect it to avoid object store and tell me if it was or wasn't able to do so rather than telling me if I can do it and on what platform am I running.

I don't know if this is a difference in dialect, but to me this renaming is much clearer and much more intuitive. Being more explicit would be awkward, all I can think of is something like `ShouldAvoidObjectStore()`, which I don't really like.

**Comment by jarrpa** [comment]
@crombus Thank you for the updates! You're going to hate me for this, but I've had a change of heart. :pray: 

In reviewing a separate PR, I decided that it would make more sense to not return the platform with this new function. My reasoning is that since the value is *only* used for two log messages, we can just make the GetPlatform() call right before the log message. In addition, I saw that there are points in the unit tests where avoidObjectStore() is called that we should also probably update.

File: controllers/storagecluster/cephblockpools_test.go
Comment (jarrpa) [inline controllers/storagecluster/cephblockpools_test.go:37]: I'm not quite sure I understand the need for this change... how does it relate to the platform detection changes? Also, why are you running `initStorageClusterResourceCreateUpdateTestWithPlatform()` twice? That seems like a waste. Either you run it with or without the new objects. It seems you should just change the one line to add the new parameter.

File: controllers/storagecluster/cephobjectstores_test.go
Comment (jarrpa) [inline controllers/storagecluster/cephobjectstores_test.go:53]: You are still ignoring these errors. Don't. Check them and fail as appropriate.

File: controllers/storagecluster/initialization_reconciler_test.go
Comment (jarrpa) [inline controllers/storagecluster/initialization_reconciler_test.go:105]: Combine all these into one `if` stanza. They all check the same thing with no logic in between.

**Review by jarrpa — CHANGES_REQUESTED** [review]
The more I look at this, the more I start to frown at the whole Platform implementation... but that refactor is a separate PR. For now I think you're on the right track within the limitations of what we have. Still a few comments to resolve.

**Comment by umangachapagain** [comment]
> For the purposes of determining if we will deploy RGW, we have no need to know what the platform actually is. This method is way more streamlined. One function call to get both the boolean and platform (if applicable) makes more sense to me.
>

Knowing what the platform actually is is exactly what we need to determine if we deploy RGW. And if we need to know it, why not make it explicit?

I'd be ok with this refactor if we  weren't doing
```
if avoid {
		platform, err := r.platform.GetPlatform(r.Client)
		...
		r.Log.Info(fmt.Sprintf("not creating a CephObjectStore because the platform is '%s'", platform))
		...
	}
```
right after the new wrapper returns.

My suggestion: `PlatformsShouldAvoidObjectStore` returns Platform, Bool, Error. Use the returned Platform for logging instead of fetching it again. This also ensures that you log the `Platform` against which you did the `avoidObjectStore` check and not your own copy of it which might change or you might not be able to retrieve it?

**Comment by jarrpa** [comment]
> I'd be ok with this refactor if we weren't doing
> 
> ```
> if avoid {
> 		platform, err := r.platform.GetPlatform(r.Client)
> 		...
> 		r.Log.Info(fmt.Sprintf("not creating a CephObjectStore because the platform is '%s'", platform))
> 		...
> 	}
> ```
> 
> right after the new wrapper returns.
> 
> My suggestion: `PlatformsShouldAvoidObjectStore` returns Platform, Bool, Error. Use the returned Platform for logging instead of fetching it again. This also ensures that you log the `Platform` against which you did the `avoidObjectStore` check and not your own copy of it which might change or you might not be able to retrieve it?

This is part of why I want to redo the entire platform detection implementation. We shouldn't be trying to both discover *and* retrieve the platform type in the same function. Even now we only ever check it once per instance of ocs-operator, we store it in the reconciler once we know it. IMO we should have `r.DetectPlatform()`, which runs once at the beginning, and `r.GetPlatform()` everywhere else. What's the likelihood that the platform OCP is running on changes *while the operator is still running*?

File: controllers/storagecluster/initialization_reconciler_test.go
Comment (jarrpa) [inline controllers/storagecluster/initialization_reconciler_test.go:75]:
```suggestion
	avoid, avoidErr := reconciler.PlatformsShouldAvoidObjectStore()
	assert.NoError(t, avoidErr)
```

File: controllers/storagecluster/storageclasses_test.go
Comment (jarrpa) [inline controllers/storagecluster/storageclasses_test.go:46]:
```suggestion
	avoid, avoidErr := reconciler.PlatformsShouldAvoidObjectStore()
	assert.NoError(t, avoidErr)
```

**Review by jarrpa — CHANGES_REQUESTED** [review]
Two final things to change in the tests. We really should not be ignoring returned errors.

File: controllers/storagecluster/storageclasses_test.go
Comment (crombus) [reply controllers/storagecluster/storageclasses_test.go]: I might have missed these. updated now?

**Review by jarrpa — APPROVED** [review]
Thank you for your extended patience!

I'd like to at least mention the desire for two subsequent changes after this merges:
1. Refactor the while `GetPlatform()` thing so that we have an explicit `DetectPlatform()`/`SetPlatform()` function we call at the beginning of reconciliation. That way we can stop passing in a client we won't be using.
2. I thought of a slight optimization that would follow after this (maybe just as another commit in the same PR) that may make this look neater:
   ```
  	if avoid {
  		r.Log.Info(fmt.Sprintf("not creating a Ceph RGW route because the platform is '%s'", r.GetPlatform()))
  		return nil
  	}
   ```
   We should figure out a way to record this somewhere more official, but I'll leave that as an offline exercise.

/lgtm

File: controllers/storagecluster/storageclasses_test.go
Comment (jarrpa) [reply controllers/storagecluster/storageclasses_test.go]: yup!

**Review by umangachapagain — APPROVED** [review]

**Pull request merged**
